feat(Sales Invoice): make validation configurable (IBT-86)

Kamillendampf opened the pull request:
Added support for configurable validation behavior with two kinds of settings:

- Turn validation on or off, separately for save and submit events
- Add an option to show either a warning or an error when the validation runs and fails

Other related work:

- Add link to new settings page to workspace
- Update German translations
- Patch to add default settings so that the behavior remains unchanged unless the user adjust the settings

Resolves #159

Kamillendampf committed:
feat(Sales Invoice): make validation configurable (IBT-86)

Added support for configurable validation behavior with three modes:
- Silen validation : validations run, but no messages
- No validation: disables all validations
- Warning: validations run, but failures only trigger warnings
- Error: validations run, and failures block further actions

This allows more flexible handling of validation logic during form submission or data insertion.

In case the user haven't set a value in E Invoice Settings, the system gave the user a warning that the default value (Silent Validation) is used.

resolves: #159

PatrickDEissler commented on eu_einvoice/hooks.py:
```diff
@@ -143,7 +143,8 @@
 	"Sales Invoice": {
-		"validate": "eu_einvoice.european_e_invoice.custom.sales_invoice.validate_doc",
+		"before_submit": "eu_einvoice.european_e_invoice.custom.sales_invoice.validate_doc",
+		"before_save": "eu_einvoice.european_e_invoice.custom.sales_invoice.validate_doc",
```
```suggestion
		"before_submit": "eu_einvoice.european_e_invoice.custom.sales_invoice.before_submit",
		"before_save": "eu_einvoice.european_e_invoice.custom.sales_invoice.before_save",
```
That is at least my personal preference, to have the run-methods extendable with further functions.
You would then call your validate_doc method in these run methods.

Example:
before_save(doc, event):
-> validate_doc(doc, event)
before_submit(doc, event):
-> validate_doc(doc, event)

Other thing: Best practice is also to sort the run methods sequentially.

PatrickDEissler commented on eu_einvoice/european_e_invoice/custom/sales_invoice.py:
```diff
@@ -744,10 +744,18 @@ def validate_doc(doc, event):
+	if event == "before_save":
+		validation_state = frappe.get_single_value("E Invoice Settings", "validation_on_save_insert")
+	if event == "before_submit":
+		validation_state = frappe.get_single_value("E Invoice Settings", "validation_on_submit")
```
```suggestion
	if event not in ["before_save", "before_submit"]:
		return
	settings_field = "validation_on_save_insert" if event == "before_save" else "validation_on_submit"
	validation_state = frappe.get_single_value("E Invoice Settings", settings_field)
```
Little refactoring (just suggestion).

PatrickDEissler commented on eu_einvoice/european_e_invoice/custom/sales_invoice.py:
```diff
@@ -774,11 +782,17 @@ def validate_einvoice(doc: SalesInvoice):
 	if any(validation_errors):
 		doc.validation_errors += "\n".join(validation_errors)
+		if validation_state == "Error":
+			frappe.throw(_("E Invoice is not correct."))
+		elif validation_state == "Warning":
+			frappe.msgprint(_("E Invoice is not correct."), alert=True, indicator="red")
 	else:
 		doc.einvoice_is_correct = 1
 
 	if any(warnings):
 		doc.validation_warnings += "\n".join(warnings)
+		if validation_state == "Warning":
+			frappe.msgprint(_("E Invoice is not correct.", warnings), alert=True, indicator="orange")
```
Maybe with the prints and throws you could mention further steps (for example: Where can the user get furhter information? Are there common suggestions to solve the problem?)
-> Not sure if this makes sense in this business context.

barredterra committed:
fix: add default value for settings

barredterra committed:
refactor: Sales Invoice validation

barredterra committed:
chore: update german translations

barredterra committed:
fix: add Settings to Workspace

barredterra committed:
refactor: remove unused event param

barredterra committed:
fix: add patch for default settings

barredterra commented:
Todo: handle exceptions

barredterra committed:
refactor: handle exceptions

Move new logic out of `validate_einvoice` so we can handle all results in a single place. Let the settings controller decide what needs to run.

barredterra committed:
fix: show message only if validation ran

barredterra committed:
docs: add section to README

barredterra commented:
@Mergifyio backport version-15 version-14
